File: Medium/SpiralMatrix.js
/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
var spiralOrder = function(matrix) {
    
  let res = [];
  let top = 0;
  let bottom = matrix.length;
  let left = 0;
  let right = matrix[0].length;
  
  while (left < right && top < bottom) {
    // top
    for (let i = left; i < right; i++)
      res.push(matrix[top][i]);
    top += 1;
    
    // right
    for (let i = top; i < bottom; i++)
      res.push(matrix[i][right - 1]);
    right -= 1;
    
    if (left >= right || top >= bottom)
      break;
    
    // bottom
    for (let i = right - 1; i >= left; i--)
      res.push(matrix[bottom - 1][i]);
    bottom -= 1;
    
    // left
    for (let i = bottom - 1; i >= top; i--)
      res.push(matrix[i][left]);
    left += 1;
  }
  
  return res;
};

/*

  Time: O(m*n)
  Space: O(1)
  
  Edge cases:
  * A matrix with a single column
  * A matrix with a single row

*/

File: C#/Medium/SpiralMatrix.cs
public class Solution
{
    public IList<int> SpiralOrder(int[][] matrix)
    {
        int nRows = matrix.Length;
        int nCols = matrix[0].Length;
        
        int top = 0;
        int bottom = nRows;
        int left = 0;
        int right = nCols;

        var result = new List<int>();
        while (top < bottom && left < right)
        {
            // Right
            for (int i = left; i < right; i++)
            {
                result.Add(matrix[top][i]);
            }
            top++;

            // Down
            for (int i = top; i < bottom; i++)
            {
                result.Add(matrix[i][right - 1]);
            }
            right--;

            if (top >= bottom || left >= right) 
            {
                break;
            }

            // Left
            for (int i = right - 1; i >= left; i--)
            {
                result.Add(matrix[bottom - 1][i]);
            }
            bottom--;

            // Up
            for (int i = bottom - 1; i >= top; i--)
            {
                result.Add(matrix[i][left]);
            }
            left++;
        }

        return result;
    }
}

/*

    Solution:
    * Keep pointers at the bounds of the matrix
    * Change directions once a boundary is reached
    * Move the boundaries towards the center of the matrix after each row/col is traversed

    Time: O(m*n)
    Space: O(1)

*/
